### DISCUSSION TITLE
Feature request: relay mode

### DISCUSSION THREAD

**Discussion by dwmw2** [gekkekoe/esphome-ecodan-hp#30] [General]
It would be really useful to be able to keep the Melcloud wifi adapter connected while also using esphome.

If we have two UARTs available, can't we relay commands from the wifi adapter to the heat pump, and relay responses (or *all* messages?) back?

**Comment by gekkekoe** [comment]
I actually use this setup to reverse engineer those missing cmds (proxy between master and a slave).
There's already a discussion about this topic here https://github.com/gekkekoe/esphome-ecodan-hp/discussions/20

I also think that except for the MMSP incentive in the UK, there's no compelling reason to keep the melcloud adapter when running esphome? But you might be able to provide the data from HA instead of melcloud? You would also need some extra hardware and do some soldering. A custom pcb can be made, but that takes some time and rework the software for it. It a lot of work and there's no gain for most users. Probably there's not enough demand for this to make it viable.

**Comment by dwmw2** [comment]
Soldering? I'm just using the recommended [PAP-05V leads](https://www.aliexpress.com/item/1005002904897793.html) from AliExpress which will plug into headers on fairly much any esp32 board. I think I've found some [leads](https://www.aliexpress.com/item/1005007176563512.html) running to a matching socket, which can just plug into other pins on the board for the second uart feeding the Melcloud adapter.

I'm not averse to soldering if I have to, but I'm also fundamentally lazy :)

I have just had a cascade system installed: two heat pumps and a master controller, each of which has a CN105 to play with. I've not heard of MMSP (is that still something I should look at?) but the installer is using Melcloud to monitor the system, so I'd like to keep it connected.

For the two units I'll probably install esphome-ecodan-hp on these Faikin modules: https://www.amazon.co.uk/dp/B0C2ZYXNYQ
I'll use something with more available pins for the one I want to relay, of course.

Once I get the hardware set up I'll see if I can make it an optional feature which isn't too intrusive. It ought to be as simple as a config option for the `proxy_uart` which refers to any configured [UART](https://esphome.io/components/uart.html).

**Comment by gekkekoe** [comment]
issue is that when I did reverse engineering, the procon did not power from the 5v only. Probably the 5v is also not enough to power 2 boards. So you need to use the 12v from the cn105 and put in a levelshifter. But thinking about it, you might be able to get away with redirecting the 12v to the slave and use the 5v for the esp32. If that works (please test, be careful not to fry your mainboard), I can cleanup the reverse engineering code to make an slave optional.

double check if https://www.amazon.co.uk/dp/B0C2ZYXNYQ is 5v tolerant, else you need to do levelshifting to 3.3v.
the m5stack s3 lite has 2 pairs of 5v tolerant pins. You can use 1+2 and 38+39 (that's what I did) see https://docs.m5stack.com/en/core/AtomS3%20Lite

**Reply by dwmw2** [reply]
It's not just "5v tolerant"; it's *designed* for precisely this use case and I think it even has pull-ups on the TX output too.

https://github.com/revk/ESP32-Faikin/tree/main/PCB

It can be powered from the 12v line (accepts 4-36V), so I think it should be good.

I have a cascade system — two heat pumps and the main controller. So I'll put the Faikin units on the two heat pumps. Might think about something with a more easily accessible second serial port for the main controller where I want to do the relaying (or just use the serial port from the debug header and actually get the soldering iron out)

**Reply by dwmw2** [reply]
Hm, I see the AtomS3 Lite does have some extra I/O pins accessible. Are they *also* 5v-tolerant? I can't find a full data sheet or schematic which even confirms that G1/G2 are. If I throw in 5v buck converter and feed it from the 12v line, maybe I'll use an AtomS3 Lite on the master unit proxying to the Melcloud adapter, and Faikins on the actual heat pump controllers.

**Reply by gekkekoe** [reply]
https://static-cdn.m5stack.com/resource/docs/products/core/AtomS3%20Lite/img-4061fdd4-6954-4709-a7e7-b0f50e5ba52e.webp
(notice the 5v, means those pin sequence are 5v).  So 1+2 and 38+39 are 5v tolerant. in the documentation you can also find it, a bit hidden. But they work, I used them both with 5v.

5,6,7,8 are 3.3v 

They are using this https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/atombase/ATOMS3/sy8089.pdf step down regulator for the 5v to 3.3v

**Reply by dwmw2** [reply]
![img-4061fdd4-6954-4709-a7e7-b0f50e5ba52e](https://github.com/user-attachments/assets/cf90d720-399c-4229-952c-bdf7ec79df36)

I hadn't interpreted it like that; I just thought that meant there happened to be a 5v pin on the right hand side in the same row as G38/G39, and there happened to be a 5v pin in the Grove connector. And a 3.3v pin on the left hand side. I'm not sure why that implies anything about the tolerances of the GPIO pins that happen to be physically close? But if you say you tested it, that works. I've ordered one; we'll see if it blows up :)

Thanks!

**Reply by gekkekoe** [reply]
![image](https://github.com/user-attachments/assets/35652f7e-6bc9-4352-bbb9-48d473ec3953)
It works for sure, see img. I did find the schematic, but it seems to be gone now, but it clearly has the stepdown for 1+2 and 38+39

**Reply by dwmw2** [reply]
FWIW support@m5stack.com responded to a query:

>> Hello, please could you tell me if the GPIOs on the AtomS3 Lite are 5v tolerant?
>> The documentation doesn't seem to say, or give any hint why the G3[89] pins are grey while the  G[125678] are green.
>
> Hi David,
>
> Good day!
>
> All IOs are 3.3V supported.

**Reply by gekkekoe** [reply]
that's is really odd (must have been an intern). even on their site https://docs.m5stack.com/en/core/AtomS3%20Lite they mentions 5v to 3.3v conversion. and why would they have a https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/atombase/ATOMS3/sy8089.pdf on board :D  I and a lot of users are using this hardware setup already. 

try to get the schematic ...

**Reply by dwmw2** [reply]
The regulator is to feed 3.3v to the chip as its power supply, from an incoming 5v feed.

That's different to tolerating 5v on the actual I/O pins. For that you need over-voltage protection on each pin too, like the ones in the Faikin:
![Screenshot from 2024-08-05 18-15-36](https://github.com/user-attachments/assets/b33c9542-b01c-4153-b7e0-44cdbd3e2e13)
See the TVS diode and FET on each line for protection, and the pull-up of the TX line to 5V (which is on pin1 of J5).

**Reply by gekkekoe** [reply]
too bad I cannot find the schematics anymore (cause I did see it, else I would not connect it directly). but the groove port is also 5v, a lot of grove devices that are compatible are using 5v.

my atom should be fried by now (and a lot of users) if there's no level shifting. been using this setup for over 6m
 months already.  temps are around 39c.

**Reply by dwmw2** [reply]
I think you have to be unlucky to fry the ESP32 with 5v on its I/O pins anyway, despite it theoretically being rated only at 3.3v?

**Reply by gekkekoe** [reply]
agreed, similar to https://arduino.stackexchange.com/questions/88619/m5stack-confused-about-i2c-bus-voltage#comment202218_88621

found more refs https://community.m5stack.com/topic/1078/5v-input-concern-on-m5stickc/18
"M5Stack units have internal 3v3 regulators to reduce the 5v down to safe voltage.
I have mentioned this in my UIFLOW book with the circuit diagram of the regulator."

too bad I did not save the schematics, now I'm having doubts if I did see it correctly.
